File: 39_combination_sum/combinationSum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> path = new ArrayList<>();

        dfs(candidates, target, path, res);

        return res;
    }

    private void dfs(int[] candidates, int target, List<Integer> path, List<List<Integer>> res) {
        if (target < 0) return;
        if (target == 0) {
            res.add(path);
            return;
        }
        for (int i = 0; i < candidates.length; i++) {
            int newTarget = target - candidates[i];
            List<Integer> newPath = new ArrayList<>(path);
            newPath.add(candidates[i]);
            dfs(Arrays.copyOfRange(candidates, i, candidates.length), newTarget, newPath, res);
        }

    }
}

File: 39_combination_sum/combinationSum.ts
function combinationSum(candidates: number[], target: number): number[][] {
  const result = [];
  dfs(candidates, target, [], result);

  return result;
}

function dfs(
  nums: number[],
  target: number,
  path: number[],
  result: number[][]
) {
  if (target < 0) {
    return 0;
  }

  if (target === 0) {
    result.push(path);
    return;
  }

  for (let i = 0; i < nums.length; i++) {
    const newTarget = target - nums[i];
    const newPath = [...path, nums[i]];

    dfs(nums.slice(i), newTarget, newPath, result);
  }
}
